MM-13893: introduce file store

lieut-data opened the pull request:
#### Summary
This PR builds on https://github.com/mattermost/mattermost-server/pull/10230 -- but targetting a feature branch -- to abstract a config interface and rematerialize a concrete `FileStore` and `MemoryStore` implementations of same. A `DatabaseStore` would follow on as a separate effort.

These changes were hampered by the complexity that is `App.Config()`, `App.GetConfig()` (now `App.GetSanitizedConfig()`), `App.UpdateConfig()` and `App.SaveConfig()` and `App.PersistConfig()`. To simplify these changes, I aimed to preserve the entirety of the current functionality and avoid needless refactoring in the application layer. 

I fully intend to narrow the surface area of the config store API -- in particular eliminating `Save` (which effectively does the aforementioned `App.PersistConfig`) and making `Set` and `Patch` persist automatically -- but I'm deferring this rework to a separate PR. I had also wanted to detect and reject an attempt to modify the config in place (vs. doing a `Clone`), but again this was hampered by implicit dependence on the current behaviour.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-13893

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

lieut-data commented on api4/emoji.go:
```diff
@@ -37,11 +37,6 @@ func createEmoji(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if len(*c.App.Config().FileSettings.DriverName) == 0 {
```
I called this out in a [separate commit](https://github.com/mattermost/mattermost-server/pull/10243/commits/41cf64c5ef20d7c97c7147a843eab9787bb90a6a), but in short I moved these checks into the `App` layer as they are a config issue not a permission issue.

I couldn't "fix" the test because it's no longer possible to use `App.UpdateConfig()` to write an invalid config (followed by an `App.PersistConfig()` that would render the server unable to start next time!). In another place, I used the new `MemoryStore` which skips such checks to continue to allow the test, but I didn't want to pull this test apart too much to enable that here.

lieut-data commented on api4/file_test.go:
```diff
@@ -928,14 +928,8 @@ func TestGetFileLink(t *testing.T) {
-	}()
 	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.EnablePublicLink = true })
-	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.PublicLinkSalt = model.NewId() })
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.PublicLinkSalt = model.NewRandomString(32) })
```
The `PublicLinkSalt` failed config validation, but an easy fix this time.

lieut-data commented on app/admin.go:
```diff
@@ -168,32 +169,19 @@ func validateLdapFilter(cfg *model.Config, ldap einterfaces.LdapInterface) *mode
-
-	if err := validateLdapFilter(cfg, a.Ldap); err != nil {
+func (a *App) SaveConfig(newCfg *model.Config, sendConfigChangeClusterMessage bool) *model.AppError {
+	if err := validateLdapFilter(newCfg, a.Ldap); err != nil {
```
Note that `validateLdapFilter` now occurs before the `IsValid` check inside `configStore.Set`. As you note above, `validateLdapFilter` only requires the `Enable` and `UserFilter` from the incoming config, so validity here wasn't an issue. For the time being, I opted not to embed this check into the config store itself in the hope that I could think about it as an "external" validity vs. an "internal" one. 

I'm not 100% sold on this -- because the enterprise code will `panic` if it discovers the filter to be invalid, so it may as well be "internal". There's a lot about the config subsystem that we could stand to improve. Thoughts welcome here.

lieut-data replied:
This comment is out of date: the `validateLdapFilter` now occurs as part of the model validity checks.

lieut-data commented on app/admin.go:
```diff
@@ -168,32 +169,19 @@ func validateLdapFilter(cfg *model.Config, ldap einterfaces.LdapInterface) *mode
-	a.UpdateConfig(func(update *model.Config) {
-		*update = *cfg
-	})
 	a.PersistConfig()
```
As noted in the summary, I preserve the `PersistConfig` interface only because various tests and business logic currently assume it's possible to `UpdateConfig` and then `ReloadConfig` to recover from the backing store if you haven't called `PersistConfig` in between. I don't think this is actually needed, but I didn't want to tear that apart right now.

The `ReloadConfig` appeared to be superfluous -- and the watcher logic is now internalized to the file store.

lieut-data commented on app/config.go:
```diff
@@ -29,21 +28,15 @@ const (
-		return cfg.(*model.Config)
-	}
-	return &model.Config{}
+	return s.configStore.Get()
```
I've been thinking about making `Get()` do a clone of the config, so that it's never possible to modify the actual config without invoking `Set` or `Patch`. We call `Config()` a /lot/, so there /might/ be some noticeable overhead to doing this, at least without a more efficient `Clone`. Either way, this breaks stuff so I'm deferring that as noted in the summary.

lieut-data commented on app/config.go:
```diff
@@ -54,56 +47,33 @@ func (s *Server) UpdateConfig(f func(*model.Config)) {
-	s.config.Store(updated)
-
-	s.InvokeConfigListeners(old, updated)
+	if _, err := s.configStore.Set(updated); err != nil {
```
Notifying listeners is now a detail internal to the store.

lieut-data commented on app/config.go:
```diff
@@ -54,56 +47,33 @@ func (s *Server) UpdateConfig(f func(*model.Config)) {
-	config.SaveConfig(a.ConfigFileName(), a.Config())
-}
-
-func (s *Server) LoadConfig(configFile string) *model.AppError {
```
The concept of `LoadConfig` doesn't make sense in a world where we can source from non-files, so this functionality has been eliminated altogether.

lieut-data commented on app/config.go:
```diff
@@ -54,56 +47,33 @@ func (s *Server) UpdateConfig(f func(*model.Config)) {
 }
 
-func (s *Server) ReloadConfig() *model.AppError {
+func (s *Server) ReloadConfig() error {
```
`error` for the win!

lieut-data commented on app/config.go:
```diff
@@ -306,51 +241,6 @@ func (a *App) regenerateClientConfig() {
 	a.Srv.clientConfigHash = fmt.Sprintf("%x", md5.Sum(clientConfigJSON))
 }
 
-func (a *App) Desanitize(cfg *model.Config) {
```
Internalized to the config store.

lieut-data commented on app/config_test.go:
```diff
@@ -4,67 +4,29 @@
-		} else if *newConfig.TeamSettings.SiteName != *originalSiteName {
-			t.Fatal("new config contains incorrect site name")
-		}
+		assert.Equal(t, *originalSiteName, *oldConfig.TeamSettings.SiteName, "old config contains incorrect site name")
```
Observe that these assertions are exactly the opposite of the previous ones: the behaviour before relied on the ability to invoke `ReloadConfig()` and discard the assignment in `UpdateConfig` above, having not done a `PersistConfig`. I'm skeptical that this was the purpose of this test, given that we don't test listeners in a more "normal" way, so I opted to make this test more lifelike.

This was also the start of an attempt to eliminate the need to `PersistConfig` at all, but I ran into other difficulties that precluded making that change right now.

lieut-data commented on app/emoji.go:
```diff
@@ -31,6 +31,14 @@ const (
 )
 
 func (a *App) CreateEmoji(sessionUserId string, emoji *model.Emoji, multiPartImageData *multipart.Form) (*model.Emoji, *model.AppError) {
+	if !*a.Config().ServiceSettings.EnableCustomEmoji {
```
Relocated from `api4` earlier.

lieut-data commented on app/options.go:
```diff
@@ -38,8 +39,19 @@ func StoreOverride(override interface{}) Option {
 
 func ConfigFile(file string, watch bool) Option {
```
The various `Option` returning functions in this file are used both at startup and during tests for configuring behaviour. `ConfigFile` preserves the current interface, while `ConfigStore` is more flexible, allowing for a `MemoryStore` or possibly the `DatabaseStore` as required in the future.

lieut-data commented on app/post_test.go:
```diff
@@ -533,8 +532,7 @@ func TestMaxPostSize(t *testing.T) {
 				Srv: &Server{
-					Store:  mockStore,
-					config: atomic.Value{},
+					Store: mockStore,
```
This test didn't need to initialize `config`, which is now gone anyway.

lieut-data commented on app/server.go:
```diff
@@ -75,10 +74,6 @@ type Server struct {
 	runjobs bool
 	Jobs    *jobs.JobServer
 
-	config                 atomic.Value
```
`configStore` replaces all of this.

lieut-data commented on app/server.go:
```diff
@@ -150,11 +142,20 @@ func NewServer(options ...Option) (*Server, error) {
 
-	s.EnableConfigWatch()
+		s.configStore = configStore
+		if needsSave {
```
I'm not super happy about `needsSave`, but our current config loading model involves detecting missing settings or warning and "fixing" them without erroring out altogether. I really didn't want `Load()` (invoked through `NewFileStore`) to have implicit write semantics, so I've forced that on the caller instead to do the requisite `Save`. This, of course, conflicts with my desire to eliminate `Save` altogether and make it implicit. Perhaps it needs to stay...

lieut-data commented on build/release.mk:
```diff
@@ -33,7 +33,6 @@ package:
 
 	@# Resource directories
 	mkdir -p $(DIST_PATH)/config
-	cp -L config/README.md $(DIST_PATH)/config
```
Removed `README.md` as it /seemed/ unnecessary: https://github.com/mattermost/mattermost-server/tree/master/config.

Opinions welcome, but it's pretty weird to have that page without a `config.json` present in the source to explain it. Plus it will eventually go away with the database config. 0/5.

crspeller replied:
Need to restore this copy.

lieut-data commented on cmd/mattermost/commands/config.go:
```diff
@@ -217,37 +211,26 @@ func configSetCmdF(command *cobra.Command, args []string) error {
 	}
 
-	if err := config.ValidateLocales(app.Config()); err != nil {
+	// UpdateConfig above would have already fixed these invalid locales, but we check again
```
Invalid locale handling is... weird. I've tried to preserve the semantics for now.

lieut-data commented on config/client.go:
```diff
@@ -4,393 +4,14 @@
 package config
 
 import (
-	"bytes"
```
Most of this has been split up into other files or rendered obsolete.

lieut-data commented on config/database.go:
```diff
@@ -0,0 +1,11 @@
+type databaseStore struct {
+}
+
+func NewDatabaseStore() *databaseStore {
```
Not yet implemented, of course!

lieut-data commented on config/emitter.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// emitter enables threadsafe registration and broadcasting to configuration listeners
```
Threadsafe, yes, but the concurrency model isn't ideal. Listeners are executed synchronously (otherwise lots of code assumptions break), so I've taken care to release all locks before issuing the callbacks (otherwise we run the rusk of deadlock with the `RWLock`.)

lieut-data commented on config/file.go:
```diff
@@ -0,0 +1,323 @@
+//
+// Consideration is given to maintaining backwards compatibility when resolving the path to the
+// configuration file.
+func resolveConfigFilePath(path string) (string, error) {
```
I'd love to deprecate relative config paths altogether if we preserve config files at all come v6.0.

lieut-data commented on config/file.go:
```diff
@@ -0,0 +1,323 @@
+		})
+	}()
+
+	oldCfg := fs.config
```
Concurrently, "bad" code could still be updating the `fs.config` returned by `Get` above. I'd like to force a `Clone` in `Get` as discussed, but deferred for now.

lieut-data commented on config/file.go:
```diff
@@ -0,0 +1,323 @@
+	return oldCfg, nil
+}
+
+// Patch merges the given config with the current configuration, without updating the backing store.
```
I haven't given this a lot of thought yet, but the rough sketch for how this will work is to accept a `model.Config` on which `SetDefaults` hasn't been called, and then to ignore all `nil` fields.

lieut-data commented on config/file.go:
```diff
@@ -0,0 +1,323 @@
+	var f io.ReadCloser
+
+	f, err = os.Open(fs.path)
+	if os.IsNotExist(err) {
```
It's perfectly fine to start out without a file: the default behaviour is to initialize with the defaults. Now, we no longer require `default.json` to be located, and instead just write out the defaults as codified in `model.Config`.

lieut-data commented on config/file.go:
```diff
@@ -0,0 +1,323 @@
+		needsSave = true
+	}
+
+	fs.configLock.Lock()
```
Note that we haven't had to acquire a lock until now in this method, because we haven't read or written to anything mutable on the file store. `path` doesn't count, since it can't be modified after initialization.

lieut-data commented on config/store.go:
```diff
@@ -0,0 +1,48 @@
+	// Set replaces the current configuration in its entirety, without updating the backing store.
+	Set(*model.Config) (*model.Config, error)
+
+	// Patch merges the given config with the current configuration, without updating the backing store.
```
The thought just occurred that it might make sense to have `Set` have `Patch` semantics for any `nil` fields as such. 0/5, but that would eliminate another method, at the expense of perhaps being less clear (unless we renamed `Set` to `Merge`?)

lieut-data commented on config/unmarshal.go:
```diff
@@ -0,0 +1,206 @@
+)
+
+// newViper creates an instance of viper.Viper configured for parsing a configuration.
+func newViper(allowEnvironmentOverrides bool) *viper.Viper {
```
Most of the code in this file was pre-existing, just relocated.

lieut-data commented on config/utils.go:
```diff
@@ -0,0 +1,133 @@
+//
+// Ideally, this function would be completely internal, but it's currently exposed to allow the cli
+// to test the config change before allowing the save.
+func FixInvalidLocales(cfg *model.Config) bool {
```
I changed the signature here to make it easier to use with `fixConfig` and to avoid having weird error propagation semantics, but the gist of the original method remains untouched.

lieut-data commented on services/mailservice/mail_test.go:
```diff
@@ -21,9 +21,11 @@ import (
 
 func TestMailConnectionFromConfig(t *testing.T) {
-	cfg, _, _, err := config.LoadConfig("config.json")
+	fs, _, err := config.NewFileStore("config.json", false)
```
Needs the actual `config.json` to connect to the `SMTPServer`. Could probably have used a memory store, though, given that we override via environment variables on the CI.

lieut-data commented:
@crspeller, as discussed offline:
* moved ldap filter into config store: actually, I moved it right out of enterprise and into the `model` package as part of the existing validity checks
* eliminate the `needsSave` paradigm, and just let `Load` write to the backing store as necessary
* restored `README.md`, unmodified for now
* remove placeholders for database store and patch operations (to come in separate PRs)

I've left this against the "feature branch" for now, until https://github.com/mattermost/mattermost-server/pull/10230 is merged into `master`, after which I'll point the commits there.

lieut-data commented:
1. Waiting on https://github.com/mattermost/mattermost-server/pull/10230 to be reviewed before I can point this at `master`. @mgdelacroix, do you have the cycles to take a look at these reviews?
2. Are you referring to the utilities for finding config files, still in `utils/`? I'm game to relocate them, but at this point I'd prefer to punt that to a second PR as I'm knee deep in the database stuff already. Is that reasonable?

lieut-data commented:
Thanks @mgdelacroix, and you're welcome! Normally, there would have been fewer commits, but @crspeller and I iterated on some changes post starting the review, so there was a bit more churn than usual.

Also, heads up that I'm going to have to actually rebase this to point against `master` given the way I had originally meant to merge to a feature branch. Normally, I'd avoid a rebase, of course, but at least this way the review is done.

crspeller commented:
@lieut-data 2. No. I am refering to functions like `structToMap` that have been copied from `utils/config.go` but have not been removed from there.

lieut-data commented:
@crspeller, perhaps I'm confused, but `utils/config.go` no longer exists -- it was renamed (and reworked) to `config/config.go` as part of https://github.com/mattermost/mattermost-server/pull/10230 and then further broken up here in this PR. 

There /are/ two implementations of `structToMap`: one in cmd/mattermost/commands/utils.go, and the other that now lives in [config/unmarshal.go](https://github.com/mattermost/mattermost-server/pull/10243/files#diff-715a53dea65304cd35eb37ebd3f86613R55). They do appear to be duplicate, modulo comments, so I can look into collapsing them. I've added a note to https://mattermost.atlassian.net/browse/MM-14143 as such -- is that fair?

crspeller commented:
@lieut-data Right. I forgot about the move from the last PR. Forget it. That duplication of my example is unrelated to this effort.
